src/main/java/gumtree/spoon/builder/NodeCreator.java

```diff
@@ -70,16 +71,33 @@ private String getClassName(String simpleName) {
 			ITree variableType = builder.createNode("VARIABLE_TYPE", type.getQualifiedName());
 			variableType.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, type);
 			type.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, variableType);
-			for (CtTypeReference<?> typeArgument: type.getActualTypeArguments()) {
-				ITree arg = builder.createNode(getClassName(typeArgument.getClass().getSimpleName()), typeArgument.getQualifiedName());
-				arg.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, typeArgument);
-				typeArgument.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, arg);
-				variableType.addChild(arg);
-			}
+			computeTreeOfTypeReferences(type, variableType);
 			builder.addSiblingNode(variableType);
 		}
 	}
 
+	@Override
+	public void scanCtActualTypeContainer(CtActualTypeContainer reference) {
+		for (CtTypeReference<?> ctTypeArgument: reference.getActualTypeArguments()) {
+			ITree typeArgument = builder.createNode(getClassName(ctTypeArgument.getClass().getSimpleName()), ctTypeArgument.getQualifiedName());
+			typeArgument.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, ctTypeArgument);
+			ctTypeArgument.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, typeArgument);
+			computeTreeOfTypeReferences(ctTypeArgument, typeArgument);
+			builder.addSiblingNode(typeArgument);
+		}
+	}
+
+	/** Creates a tree of nested type references where each nested type reference is a child of its container. */
+	private void computeTreeOfTypeReferences(CtTypeReference<?> type, ITree parentType) {
+		for (CtTypeReference<?> ctTypeArgument: type.getActualTypeArguments()) {
+			ITree typeArgument = builder.createNode(getClassName(ctTypeArgument.getClass().getSimpleName()), ctTypeArgument.getQualifiedName());
+			typeArgument.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, ctTypeArgument);
+			ctTypeArgument.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, typeArgument);
```

Comment:
This segment has been repeated a lot of times in this file. Maybe we can refactor it.